fastplotlib/graphics/_collection_base.py

```diff
@@ -0,0 +1,277 @@
+        for g in self.graphics:
+            g.remove_event_handler(callback, *types)
+
+    def __getitem__(self, item):
+        return self.graphics[item]
+
+    def __len__(self):
+        return len(self._selection)
+
+    def __repr__(self):
+        return (
+            f"{self.__class__.__name__} @ {hex(id(self))}\n"
+            f"Selection of <{len(self._selection)}> {self._selection[0].__class__.__name__}"
+        )
+
+
+class GraphicCollection(Graphic):
+    """Graphic Collection base class"""
+
+    child_type: type
+    _indexer: type
+
+    def __init__(self, name: str = None):
+        super().__init__(name)
+
+        # list of mem locations of the graphics
+        self._graphics: list[str] = list()
+
+        self._graphics_changed: bool = True
+        self._graphics_array: np.ndarray[Graphic] = None
+
+    @property
+    def graphics(self) -> np.ndarray[Graphic]:
+        """The Graphics within this collection. Always returns a proxy to the Graphics."""
+        if self._graphics_changed:
+            proxies = [
+                weakref.proxy(COLLECTION_GRAPHICS[addr]) for addr in self._graphics
+            ]
+            self._graphics_array = np.array(proxies)
+            self._graphics_array.flags["WRITEABLE"] = False
+            self._graphics_changed = False
+
+        return self._graphics_array
+
+    def add_graphic(self, graphic: Graphic):
+        """
+        Add a graphic to the collection.
+
+        Parameters
+        ----------
+        graphic: Graphic
+            graphic to add, must be a real ``Graphic`` not a proxy
+
+        """
+
+        if not type(graphic) == self.child_type:
+            raise TypeError(
+                f"Can only add graphics of the same type to a collection.\n"
+                f"You can only add {self.child_type.__name__} to a {self.__class__.__name__}, "
+                f"you are trying to add a {graphic.__class__.__name__}."
+            )
+
+        addr = graphic._fpl_address
+        COLLECTION_GRAPHICS[addr] = graphic
+
+        self._graphics.append(addr)
+
+        self.world_object.add(graphic.world_object)
+
+        self._graphics_changed = True
+
+    def remove_graphic(self, graphic: Graphic):
+        """
+        Remove a graphic from the collection.
+
+        Note: Only removes the graphic from the collection. Does not remove
+        the graphic from the scene, and does not delete the graphic.
+
+        Parameters
+        ----------
+        graphic: Graphic
+            graphic to remove
+
+        """
+
+        self._graphics.remove(graphic._fpl_address)
+
+        self.world_object.remove(graphic.world_object)
+
+        self._graphics_changed = True
+
+    def add_event_handler(self, *args):
+        raise NotImplementedError("Slice graphic collection to add event handlers")
+
+    def remove_event_handler(self, callback, *types):
+        raise NotImplementedError("Slice graphic collection to remove event handlers")
```

Comment:
so if I have a `LineCollection` of circles, do I need to have an image underlayed in order to get a click event? but then I have manually map the click event location to the closest circle...

Reply:
You need to click the actual line. Or You can add a click event handler to the renderer and then map that to world space, I should add a wrapper to PlotArea so it can directly add renderer events